File: java/patterns/src/main/java/org/prototypepatterns/structural/compositeentity/CompositeEntityPattern.java
package org.prototypepatterns.structural.compositeentity;

import java.util.List;

public final class CompositeEntityPattern {
    private CompositeEntityPattern() {
    }

    static final class CoarseGrainedOrder {
        private String customerName = "";
        private String city = "";

        void setData(String customerName, String city) {
            this.customerName = customerName;
            this.city = city;
        }

        List<String> getData() {
            return List.of(customerName, city);
        }
    }

    public static final class CompositeOrderEntity {
        private final CoarseGrainedOrder order = new CoarseGrainedOrder();

        public void setOrder(String customerName, String city) {
            order.setData(customerName, city);
        }

        public List<String> readOrder() {
            return order.getData();
        }
    }

    public static List<String> demo() {
        CompositeOrderEntity entity = new CompositeOrderEntity();
        entity.setOrder("Sam", "Austin");
        return entity.readOrder();
    }
}

File: csharp/patterns/src/PrototypePatterns/Structural/CompositeEntity/CompositeEntityPattern.cs
namespace PrototypePatterns.Structural.CompositeEntity;

public sealed class CoarseGrainedOrder
{
    private string _customerName = string.Empty;
    private string _city = string.Empty;

    public void SetData(string customerName, string city)
    {
        _customerName = customerName;
        _city = city;
    }

    public IReadOnlyList<string> GetData() => [_customerName, _city];
}

public sealed class CompositeOrderEntity
{
    private readonly CoarseGrainedOrder _order = new();

    public void SetOrder(string customerName, string city) => _order.SetData(customerName, city);

    public IReadOnlyList<string> ReadOrder() => _order.GetData();
}

public static class CompositeEntityPattern
{
    public static IReadOnlyList<string> Demo()
    {
        var entity = new CompositeOrderEntity();
        entity.SetOrder("Sam", "Austin");
        return entity.ReadOrder();
    }
}
